### PR TITLE
[RISCV][TTI] Reduce cost of a build_vector pattern

### PR DESCRIPTION
**Pull request by preames**
This change is actually two related changes, but they're very hard to meaningfully separate as the second balances the first, and yet doesn't do much good on it's own.

First, we can reduce the cost of a build_vector pattern.  Our current costing for this defers to generic insertelement costing which isn't unreasonable, but also isn't correct.  While inserting N elements requires N-1 slides and N vmv.s.x, doing the full build_vector only requires N vslide1down.  (Note there are other cases that our build vector lowering can do more cheaply, this is simply the easiest upper bound which appears to be "good enough" for SLP costing purposes.)

Second, we need to tell SLP that calls don't preserve vector registers.  Without this, SLP will vectorize scalar code which performs e.g. 4 x float @exp calls as two <2 x float> @exp intrinsic calls.  Oddly, the costing works out that this is in fact the optimal choice - except that we don't actually have a <2 x float> @exp, and unroll during DAG.  This would be fine (or at least cost neutral) except that the libcall for the scalar @exp blows all vector registers.  So the net effect is we added a bunch of spills that SLP had no idea about.  Thankfully, AArch64 has a similiar problem, and has taught SLP how to reason about spill cost once the right TTI hook is implemented.

Now, for some implications...

The SLP solution for spill costing has some inaccuracies.  In particular, it basically just guesses whether a intrinsic will be lowered to a call or not, and can be wrong in both directions. It also has no mechanism to differentiate on calling convention.

This has the effect of making partial vectorization (i.e. starting in scalar) more profitable.  In practice, the major effect of this is to make it more like SLP will vectorize part of a tree in an intersecting forrest, and then vectorize the remaining tree once those uses have been removed.

This has the effect of biasing us slightly away from strided, or indexed loads during vectorization - because the scalar cost is more accurately modeled, and these instructions look relevatively less profitable.

### CODE REVIEW

File: llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp [hunk 1]
```diff
@@ -616,6 +616,39 @@ InstructionCost RISCVTTIImpl::getShuffleCost(TTI::ShuffleKind Kind,
...
+  // lowering today does not have a first class node for this pattern.
+  InstructionCost Cost = BaseT::getScalarizationOverhead(
+      Ty, DemandedElts, Insert, Extract, CostKind);
+  std::pair<InstructionCost, MVT> LT = getTypeLegalizationCost(Ty);
+  if (Insert && !Extract && LT.first.isValid() && LT.second.isVector() &&
+      Ty->getScalarSizeInBits() != 1) {
+    MVT ContainerVT = LT.second;
+    if (ContainerVT.isFixedLengthVector())
+      ContainerVT = TLI->getContainerForFixedLengthVector(ContainerVT);
+    if (isM1OrSmaller(ContainerVT)) {
```
Comment (lukel97) [inline llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp:643]: Was there a specific reason for limiting it to M1 and below? I just did a quick check and this:

```llvm
define <8 x i32> @foo(i32 %e1, i32 %e2, i32 %e3, i32 %e4, i32 %e5, i32 %e6, i32 %e7, i32 %e8) {
  %v1 = insertelement <8 x i32> poison, i32 %e1, i32 0
  %v2 = insertelement <8 x i32> %v1, i32 %e2, i32 1
  %v3 = insertelement <8 x i32> %v2, i32 %e3, i32 2
  %v4 = insertelement <8 x i32> %v3, i32 %e4, i32 3
  %v5 = insertelement <8 x i32> %v4, i32 %e5, i32 4
  %v6 = insertelement <8 x i32> %v5, i32 %e6, i32 5
  %v7 = insertelement <8 x i32> %v6, i32 %e7, i32 6
  %v8 = insertelement <8 x i32> %v7, i32 %e8, i32 7
  ret <8 x i32> %v8
}
```

Gets lowered to a bunch of M2 vslidedowns

```nasm
buildvec_v8i32_pack: 
	vsetivli	zero, 8, e32, m2, ta, ma
	vmv.v.x	v8, a0
	vslide1down.vx	v8, v8, a1
	vslide1down.vx	v8, v8, a2
	vslide1down.vx	v8, v8, a3
	vslide1down.vx	v8, v8, a4
	vslide1down.vx	v8, v8, a5
	vslide1down.vx	v8, v8, a6
	vslide1down.vx	v8, v8, a7
	ret
```

Although it looks like with higher LMULs it ends up going through the stack at some point.

File: llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp [hunk 2]
```diff
@@ -616,6 +616,39 @@ InstructionCost RISCVTTIImpl::getShuffleCost(TTI::ShuffleKind Kind,
...
+  // worse than one vslide1down.vx per element in the type.  We could
+  // in theory do an explode_vector in the inverse manner, but our
+  // lowering today does not have a first class node for this pattern.
+  InstructionCost Cost = BaseT::getScalarizationOverhead(
+      Ty, DemandedElts, Insert, Extract, CostKind);
+  std::pair<InstructionCost, MVT> LT = getTypeLegalizationCost(Ty);
+  if (Insert && !Extract && LT.first.isValid() && LT.second.isVector() &&
+      Ty->getScalarSizeInBits() != 1) {
+    MVT ContainerVT = LT.second;
+    if (ContainerVT.isFixedLengthVector())
```
[the remark this answers is not in the document]
Comment (preames) [reply llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp:641]: Not if this API were called from the LoopVectorizer.

[the remark this answers is not in the document]
Comment (lukel97) [reply llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp:641]: We're bailing on scalable vectors at the start of the function, and we cast Ty to a fixed length vector below. Fixed length vectors are legal so surely getTypeLegalizationCost won't legalize a fixed length to a scalable?

File: llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp (hunk 1 above)
Comment (lukel97) [reply llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp:643]: Starting with a subset in this patch is fine by me. If we do want to extend this further down the line we could use getRISCVInstructionCost(RISCV::VSLIDE1DOWN_VX, ContainerVT, CostKind)

**Review by lukel97 — COMMENTED** [review]
I'm seeing a 0.88% regression on 511.povray_r on the BPI F3 after applying this, and it's fairly reproducible (< 0.1% stddev). Looking through the codegen changes, it looks like we're now avoiding partial vectorization in some places where we e.g. exploded a vector to do a bunch of exp intrinsic calls, spilling the vector registers exactly as you described. 

I would have thought that avoiding these vector spills would have been more performant, I'm not sure why it's turning out to be slower in the scalar form. Do we need discount build_vectors a bit more to get it to partially vectorize these parts again?

To clarify, I think the changes in this PR are the right thing to do, I just want to point the interaction with SLP. I'm running the other benchmarks now to see if they're also affected.

**Comment by preames** [comment]
> I'm seeing a 0.88% regression on 511.povray_r on the BPI F3 after applying this, and it's fairly reproducible (< 0.1% stddev). Looking through the codegen changes, it looks like we're now avoiding partial vectorization in some places where we e.g. exploded a vector to do a bunch of exp intrinsic calls, spilling the vector registers exactly as you described.

Just to confirm, you're looking at cycle count right?  What routine are you seeing this in?  I'm looking at an LTO build of povray, and not seeing any heavy use of the @exp routine - except indirectly through a function pointer table.  Is your build -Ofast -flto=auto?  Or something else?

> I would have thought that avoiding these vector spills would have been more performant, I'm not sure why it's turning out to be slower in the scalar form. Do we need discount build_vectors a bit more to get it to partially vectorize these parts again?

Either that, or we exposing some other issue.  If you can show me the relevant areas, I'll try to take a look.  

> To clarify, I think the changes in this PR are the right thing to do, I just want to point the interaction with SLP. I'm running the other benchmarks now to see if they're also affected.

Ack.

**Comment by lukel97** [comment]
> Just to confirm, you're looking at cycle count right? What routine are you seeing this in? I'm looking at an LTO build of povray, and not seeing any heavy use of the @exp routine - except indirectly through a function pointer table. Is your build -Ofast -flto=auto? Or something else?

This build was with -O3 -mcpu=spacemit-x60, I'll queue up another run with -Ofast and LTO.

One example I found was in `pov::compute_backtrace_texture(float*, pov::Texture_Struct*, double*, double*, pov::Ray_Struct*, double, pov::istk_entry*) (_ZN3povL25compute_backtrace_textureEPfPNS_14Texture_StructEPdS3_PNS_10Ray_StructEdPNS_10istk_entryE)`

```diff
-       flw     fa5, 36(s1)
-       fld     fs3, %pcrel_lo(.Lpcrel_hi250)(a1)
-       fld     fs4, 0(s10)
-       fcvt.d.s        fa5, fa5
-       fsub.d  fa5, fs3, fa5
-       fneg.d  fa5, fa5
-       fmul.d  fa5, fs4, fa5
-       fdiv.d  fa0, fa5, fs2
-       call    exp
-       flw     fa5, 40(s1)
-       fmv.d   fs1, fa0
-       fcvt.d.s        fa5, fa5
-       fsub.d  fa5, fs3, fa5
-       fneg.d  fa5, fa5
-       fmul.d  fa5, fs4, fa5
-       fdiv.d  fa0, fa5, fs2
-       call    exp
-       vsetivli        zero, 2, e64, m1, ta, ma
-       vfmv.v.f        v8, fs1
-       flw     fa5, 44(s1)
-       vfslide1down.vf v8, v8, fa0
-       vfmul.vf        v8, v8, fs0
-       vsetvli zero, zero, e32, mf2, ta, ma
-       vfncvt.f.f.w    v9, v8
-       csrr    a0, vlenb
-       add     a0, a0, sp
-       addi    a0, a0, 2047
-       addi    a0, a0, 65
-       vs1r.v  v9, (a0)                        # Unknown-size Folded Spill
-       fcvt.d.s        fa5, fa5
-       fsub.d  fa5, fs3, fa5
-       fneg.d  fa5, fa5
-       fmul.d  fa5, fs4, fa5
-       fdiv.d  fa0, fa5, fs2
-       call    exp
-       csrr    a0, vlenb
-       add     a0, a0, sp
-       addi    a0, a0, 2047
-       addi    a0, a0, 65
-       vl1r.v  v9, (a0)                        # Unknown-size Folded Reload
-       fmul.d  fa5, fa0, fs0
-       fcvt.s.d        fs0, fa5
-       vsetivli        zero, 2, e32, mf2, ta, ma
+       flw     fa5, 36(s1)
+       fld     fs4, %pcrel_lo(.Lpcrel_hi250)(s0)
+       fld     fs5, 0(s10)
+       fcvt.d.s        fa5, fa5
+       fsub.d  fa5, fs4, fa5
+       fneg.d  fa5, fa5
+       fmul.d  fa5, fs5, fa5
+       fdiv.d  fa0, fa5, fs2
+       call    exp
+       flw     fa5, 40(s1)
+       fmul.d  fa4, fa0, fs1
+       fcvt.s.d        fs0, fa4
+       fcvt.d.s        fa5, fa5
+       fsub.d  fa5, fs4, fa5
+       fneg.d  fa5, fa5
+       fmul.d  fa5, fs5, fa5
+       fdiv.d  fa0, fa5, fs2
+       call    exp
+       flw     fa5, 44(s1)
+       fmul.d  fa4, fa0, fs1
+       fcvt.s.d        fs3, fa4
+       fcvt.d.s        fa5, fa5
+       fsub.d  fa5, fs4, fa5
+       fneg.d  fa5, fa5
+       fmul.d  fa5, fs5, fa5
+       fdiv.d  fa0, fa5, fs2
+       call    exp
+       fmul.d  fa5, fa0, fs1
```

But at the same time, in `pov::do_light(pov::Light_Source_Struct*, double*, pov::Ray_Struct*, pov::Ray_Struct*, double*, float*) (_ZN3povL8do_lightEPNS_19Light_Source_StructEPdPNS_10Ray_StructES4_S2_Pf)` we actually go in the other direction

```diff
-       fneg.d  fa5, fa5
-       fsd     fa5, 24(s0)
-       fneg.d  fa5, fa4
-       fsd     fa5, 32(s0)
+       vsetivli        zero, 2, e64, m1, ta, ma
+       vfmv.v.f        v8, fa5
+       vfslide1down.vf v8, v8, fa4
+       vfneg.v v8, v8
+       vse64.v v8, (s5)
```

Unfortunately none of the hot *_Intersection methods seem to be affected, instead it's a large number of cold functions that are slightly perturbed. 

I'm really not sure how to interpret these changes. If the rest of the SPEC benchmarks are OK, I would be fine just chalking this up to SLP "noise".

**Comment by lukel97** [comment]
Just a quick update on the benchmarking, with LTO the regression disappears. 511.povray_r is actually significantly faster with LTO, by about 5.5%. I'm rerunning the benchmarks again with LTO, but not with -Ofast. -ffast-math seems to cause the tests to fail

**Comment by preames** [comment]
> One example I found was in `pov::compute_backtrace_texture(float*, pov::Texture_Struct*, double*, double*, pov::Ray_Struct*, double, pov::istk_entry*) (_ZN3povL25compute_backtrace_textureEPfPNS_14Texture_StructEPdS3_PNS_10Ray_StructEdPNS_10istk_entryE)`

I spent some time looking at this routine yesterday, and am still digging today.  I've been focusing less on what the current codegen looks like, than what optimal code for this sequence might look like.  (Mostly because I was looking at a differently compiled binary.)  I see a couple of possible improvements which would help both before and after significantly, but none of them are easy.  I'm going to switch and look more closely at the regression today.

**Comment by lukel97** [comment]
Here's the benchmark diffs with LTO: https://lnt.lukelau.me/db_default/v4/nts/8?show_delta=yes&show_previous=yes&show_stddev=yes&show_mad=yes&show_all=yes&show_all_samples=yes&show_sample_counts=yes&show_small_diff=yes&num_comparison_runs=0&test_filter=&test_min_value_filter=&aggregation_fn=min&MW_confidence_lv=0.05&compare_to=7

The povray regression is gone, the only noticeable change left is a 2.08% regression in 541.leela_r. In the hottest function `FastState::play_random_move()` there's a few more places where we now vectorize to a vredor.vs:

```asm
	vsetivli	zero, 0x1, e64, m1, ta, ma
	vmv.s.x	v8, a2
      	vsetivli	zero, 0x4, e16, mf4, ta, ma
	vmseq.vi	v0, v8, 0x1
      	slliw	a2, t0, 0x8
      	vsetvli	zero, zero, e32, mf2, ta, mu
      	vmv.v.i	v8, 0x0
      	ld	a4, 0x38(sp)
      	vle32.v	v8, (a4), v0.t
      	slliw	a4, a6, 0x6
	slliw	a5, t1, 0x4
      	slliw	a3, a3, 0x2
	vredor.vs	v8, v8, v8
	vmv.x.s	s1, v8
```

We're doing a packed e16 build vector which is then converted to a mask vector. 
I think we can avoid the vmv.v.i if we move the masking from the vle32.v to the vredor.vs. But that's orthogonal to this PR and I don't think the vectorized code is bad per say, so I think this is fine.

**Review by lukel97 — APPROVED** [review]
Modulo the ContainerVT nit LGTM

**Comment by preames** [comment]
> We're doing a packed e16 build vector which is then converted to a mask vector.

Not quite.  I found the surrounding code, and I'm pretty sure what we originally had was a build vector of variable i1, which was *lowered to* an i16 build vector and a comparison against zero.  Which is quite odd, as this patch shouldn't be impacting the costing of a <N x i1> build_vector.

> I think we can avoid the vmv.v.i if we move the masking from the vle32.v to the vredor.vs. But that's orthogonal to this PR and I don't think the vectorized code is bad per say, so I think this is fine.

I agree on both points.  I would assume the original code had a vselect with the neutral value after the load and before the reduction.

File: llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp [hunk 3]
```diff
@@ -616,6 +616,39 @@ InstructionCost RISCVTTIImpl::getShuffleCost(TTI::ShuffleKind Kind,
...
+      ContainerVT = TLI->getContainerForFixedLengthVector(ContainerVT);
+    if (isM1OrSmaller(ContainerVT)) {
+      InstructionCost BV = cast<FixedVectorType>(Ty)->getNumElements();
```
Comment (alexey-bataev) [inline llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp:644]:
```suggestion
      InstructionCost BV = cast<FixedVectorType>(Ty)->getNumElements() * getRISCVInstructionCost(RISCV::VSLIDE1DOWN_VX, LT.second, CostKind);
```
?

**Comment by preames** [comment]
> > We're doing a packed e16 build vector which is then converted to a mask vector.
> 
> Not quite. I found the surrounding code, and I'm pretty sure what we originally had was a build vector of variable i1, which was _lowered to_ an i16 build vector and a comparison against zero. Which is quite odd, as this patch shouldn't be impacting the costing of a build_vector.

Just noting that I also see a large number of instances of this pattern in leela *without* this patch.  Looking at them, I'm less sure if this is lowering, or a straight forward vectorization of the original code.  It does look like something we could improve both from the perspective of the i1 vector creation, and that most of the masks are indexed loads (which aren't being vectorized at all).

**Comment by preames** [comment]
I figured out the root cause of the Leela thing, and filed a ticket so we don't forget.  This is much sillier than we'd first realized.  Vectorizing this is a quite terrible result, but seems to have little to do with this patch.  https://github.com/llvm/llvm-project/issues/109466

**Pull request merged**